### PR TITLE
ToolInfo - use static dir check for composer discovery

### PR DESCRIPTION
**Pull request by Slamdunk**
In a Composer installation `php-cs-fixer` binary is present in
* `vendor/friendsofphp/php-cs-fixer/php-cs-fixer`
* `vendor/bin/php-cs-fixer` (symlink)

but someone may want to re-symlink the symlink in a custom/private bin dir.
If so, the script currently fails to detect whether is installed through composer or not.

This is due to the `$_SERVER['SCRIPT_NAME']` variable, that could vary a lot.
But composer installations can have only 2 result path, so just check them statically.

### CODE REVIEW

File: tests/ToolInfoTest.php [hunk 1]
```diff
@@ -23,11 +23,61 @@
...
+     */
+    public function testIsInstalledByComposer($scriptName)
+    {
+        $phpunitBin = $this->getRootDir().'/vendor/bin/phpunit';
+        if (!is_file($phpunitBin)) {
+            $this->markTestSkipped(sprintf('Binary "%s" is necessary to emulate PHP-CS-Fixer installed through composer', $phpunitBin));
+        }
+
+        if (!is_link($scriptName)) {
+            $this->markTestSkipped(sprintf('OS "%s" doesn\'t support the way the "%s" symlink is written', PHP_OS, $scriptName));
```
Comment (Slamdunk) [inline tests/ToolInfoTest.php]: I don't like very much the need to skip Windows, but neither I like to create symlink at runtime.
As of yet the two symlink are created under `tests/Fixtures` and committed.

File: src/ToolInfo.php [hunk 2]
```diff
@@ -26,67 +26,89 @@
...
+     * @var null|bool
+     */
+    private $isInstalledByComposer;
+
+    /**
+     * @var null|string
+     */
+    private $scriptDir;
+
+    public function __construct($script = null)
```
Comment (keradus) [inline src/ToolInfo.php]: please add typehinting as `?string`, no need for default value IMO

File: tests/ToolInfoTest.php [hunk 3]
```diff
@@ -23,11 +23,61 @@
...
     {
-        $this->assertFalse(ToolInfo::isInstalledAsPhar());
+        $toolInfo = new ToolInfo();
+        $this->assertFalse($toolInfo->isInstalledAsPhar());
+    }
+
+    /**
+     * @dataProvider provideScriptName
+     *
+     * @param mixed $scriptName
```
Comment (keradus) [inline tests/ToolInfoTest.php]: please don't use `mixed`

**Comment by keradus** [comment]
I like the idea.
Anyone willing to review this PR ?

File: src/ToolInfo.php (hunk 2 above)
Comment (Slamdunk) [inline src/ToolInfo.php]: So the code that uses `ToolInfo` should specify everytime `$_SERVER['SCRIPT_NAME']`?
Morover, it's a bit early to use 7.1-only feature, isn't it?

Comment (keradus) [inline src/ToolInfo.php]: with `?string`, one could still use `new ToolInfo(null)`, which would means exactly the same as `new ToolInfo()` with null in constructor prototype.

let me rephrase, please put typehinting for parameter in docblock, so one won't provide here an instance of `SplFile` or sth...

File: src/ToolInfo.php [hunk 4]
```diff
@@ -26,67 +26,92 @@
...
+    /**
+     * @var null|string
+     */
+    private $scriptDir;
+
+    /**
+     * @param ?string $script
+     */
+    public function __construct($script = null)
     {
```
Comment (Slamdunk) [inline src/ToolInfo.php]: @keradus is this ok?

Comment (keradus) [inline src/ToolInfo.php]: oh, ok, let it be. I hoped for non-optional parameter

File: tests/ToolInfoTest.php (hunk 1 above)
Comment (keradus) [inline tests/ToolInfoTest.php]: argh, I totally cannot work with two created symlink files on my OS (Win10)

File: src/ToolInfo.php (hunk 4 above)
Comment (Slamdunk) [inline src/ToolInfo.php]: I prefer non-optional parameter too, but only in a well-organized dependency injection context.
As of yet, ToolInfo needs to be instantiade where it is used.

Comment (keradus) [inline src/ToolInfo.php]: still, one could make it like `new ToolInfo($path)` or `new ToolInfo(null)`, right ?

File: src/ToolInfo.php [hunk 5]
```diff
@@ -26,67 +26,92 @@
...
+    /**
+     * @var null|string
+     */
+    private $scriptDir;
+
+    /**
+     * @param string $script
+     */
+    public function __construct($script)
     {
```
Comment (Slamdunk) [inline src/ToolInfo.php]: The perfect final form of this constructor, imho, would it to be `$this->script = $script;` and somewhere else a factory that injects the real script name, and factories for classes that use it that inject `ToolInfo` through its factory.
In the meantime...

File: src/ToolInfo.php [hunk 6]
```diff
@@ -26,67 +26,92 @@
...
+     */
+    private $isInstalledByComposer;
+
+    /**
+     * @var null|string
+     */
+    private $scriptDir;
+
+    /**
+     * @param string $script
```
Comment (keradus) [inline src/ToolInfo.php]: from the code I see it's nullable

File: src/ToolInfo.php (hunk 5 above)
Comment (keradus) [inline src/ToolInfo.php]: :+1: but https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2637/files#r108142119

File: src/ToolInfo.php (hunk 6 above)
Comment (Slamdunk) [inline src/ToolInfo.php]: I'm not used to phpdoc: `?string` or `null|string`?

Comment (SpacePossum) [inline src/ToolInfo.php]: the later
:+1:

Comment (keradus) [inline src/ToolInfo.php]: :+1: for earlier
both are acceptable

**Comment by keradus** [comment]
@julienfalque , could you make a review please?

File: tests/ToolInfoTest.php [hunk 7]
```diff
@@ -23,11 +23,61 @@
...
+        return array(
+            array($this->getRootDir().'/vendor/bin/phpunit'),
+            array(__DIR__.'/Fixtures/toolinfo-phpunit-symlink'),
+            array(__DIR__.'/Fixtures/toolinfo-phpunit-double-symlink'),
+        );
+    }
+
+    public function testGetComposerVersionThrowExceptionIfOutsideComposerScope()
+    {
+        $toolInfo = new ToolInfo(uniqid('non_existent_script_'));
```
Comment (julienfalque) [inline tests/ToolInfoTest.php]: `uniqid()` does not seem to be necessary here.

File: tests/ToolInfoTest.php (hunk 1 above)
Comment (SpacePossum) [inline tests/ToolInfoTest.php]: can we create and remove the symlink(s) in the setup/teardown of the tests or would that complicate the tests to much?

File: src/ToolInfo.php [hunk 8]
```diff
@@ -26,67 +26,92 @@
...
+     */
+    private $scriptDir;
+
+    /**
+     * @param null|string $script
+     */
+    public function __construct($script)
     {
-        static $result;
+        $this->script = (null !== $script ? $script : $_SERVER['SCRIPT_NAME']);
```
Comment (Slamdunk) [inline src/ToolInfo.php]: IMHO this:
```php
$this->script = null !== $script ? $script : $_SERVER['SCRIPT_NAME'];
```
Is unreadable, [operator precedence](http://php.net/manual/en/language.operators.precedence.php) can be a nightmare without parenthesis.

**Comment by Slamdunk** [comment]
> Could you also write a test for isInstalledByComposer() returning false when not installed through Composer (if possible)?

Not feasable, https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.1.2/src/ToolInfo.php#L76 will always be true in the test environment and can't be mocked without writing production code just for the purpose of mocking.

**Comment by julienfalque** [comment]
Unless I'm missing something, I think it can work with `new ToolInfo('script_outside_vendor_dir')` like you did in the last test.

File: src/ToolInfo.php (hunk 8 above)
Comment (julienfalque) [inline src/ToolInfo.php]: It's pretty obvious to me that ternary operator will be evaluated before the assignment as the opposite would almost defeat the purpose of the ternary operator, but I agree parentheses still help for readability. What do others think?

File: tests/ToolInfoTest.php [hunk 9]
```diff
@@ -21,13 +21,87 @@
  */
 final class ToolInfoTest extends \PHPUnit_Framework_TestCase
 {
+    const PHPUNIT_BIN = '/../vendor/bin/phpunit';
+
+    protected function setUp()
+    {
+        foreach ($this->provideScriptName() as $testArguments) {
+            $symlink = current($testArguments);
+            if ($symlink === self::PHPUNIT_BIN or is_link(__DIR__.$symlink)) {
```
Comment (keradus) [inline tests/ToolInfoTest.php]: oh damn, `or`...

Comment (Slamdunk) [inline tests/ToolInfoTest.php]: Will change. As of yet I'm trying to figure out how to run all on Windows.

OT: I cannot understand the love of the community for `||` and `&&`, code should be readable as fast as possible. Words will always be faster than symbols.

File: tests/Fixtures/symlinks/.gitignore [hunk 10]
```diff
@@ -0,0 +1,2 @@
+*
```
Comment (julienfalque) [inline tests/Fixtures/symlinks/.gitignore]: I would add the directory to the root .gitignore file instead, with a comment to explain its purpose.

File: tests/ToolInfoTest.php [hunk 11]
```diff
@@ -21,13 +21,87 @@
  */
 final class ToolInfoTest extends \PHPUnit_Framework_TestCase
 {
+    const PHPUNIT_BIN = '/../vendor/bin/phpunit';
+
+    protected function setUp()
```
Comment (julienfalque) [inline tests/ToolInfoTest.php]: I'm not sure using `setUp`/`tearDown` is relevant: the created symlinks are only used in `testIsInstalledByComposer`, maybe it would be easier to handle them in the test directly?

File: tests/ToolInfoTest.php [hunk 12]
```diff
@@ -21,13 +21,87 @@
...
+            array('/Fixtures/symlinks/toolinfo-phpunit-symlink'),
+            array('/Fixtures/symlinks/toolinfo-phpunit-double-symlink'),
+        );
+    }
+
+    public function testGetComposerVersionThrowExceptionIfOutsideComposerScope()
+    {
+        $toolInfo = new ToolInfo('non_existent_script');
+
+        $this->assertFalse($toolInfo->isInstalledByComposer());
```
Comment (julienfalque) [inline tests/ToolInfoTest.php]: This assertion should be moved to a new test.

File: tests/ToolInfoTest.php (hunk 9 above)
Comment (keradus) [inline tests/ToolInfoTest.php]: For the sake of homogeneity and safety.
First, as the project currently never uses `or`.
Second, as most devs cannot properly understand that difference - even if you would, then new contributor probably wont when he will be touching this code.

Comment (Slamdunk) [inline tests/ToolInfoTest.php]: Never intented to troll or change the status quo.
For sure consistency and homogeneity come first: will change in next commits.
My OT was more generic, outside this project.

File: tests/ToolInfoTest.php (hunk 11 above)
Comment (Slamdunk) [inline tests/ToolInfoTest.php]: If an assertion fails, the test stop, but `tearDown` is still called

File: tests/ToolInfoTest.php (hunk 12 above)
Comment (keradus) [inline tests/ToolInfoTest.php]: I would rename the method to `testIsInstalledAsPhar` and that's it

Comment (julienfalque) [inline tests/ToolInfoTest.php]: This method already exists, and it would not follow PHPUnit conventions IMO.

**Comment by Slamdunk** [comment]
Ok, tests now run both on Windows and Linux.
The main difference is that composer creates binaries under `vendor/bin` in different ways: on Linux is just a symlink, and on Windows it's a special executable, not a symlink, so we need to uniform the method to find `composer.json` according to this distinction.

File: src/ToolInfo.php [hunk 13]
```diff
@@ -23,83 +23,114 @@
...
-        if (null === $result) {
-            $script = $_SERVER['SCRIPT_NAME'];
+        if (null === $this->composerInstalledFile) {
+            $script = $this->script;
+
+            while (is_link($script)) {
+                // On Windows, composer creates under vendor/bin executables that
+                // aren't symlinks, so we need to manually stop at a common point
+                // equal from Windows to Linux
+                if ('vendor'.DIRECTORY_SEPARATOR.'bin' === substr(dirname($script), -10)) {
```
Comment (julienfalque) [inline src/ToolInfo.php]: AFAIK, `DIRECTORY_SEPARATOR` isn't required as paths using `/` work on Windows with PHP.

Also, this will fail when Composer is configured to install dependencies in a different directory than `vendor`.

Comment (keradus) [inline src/ToolInfo.php]: it's required as you compare it to result of dirname, which return data with OS respecting dir separator.

File: tests/ToolInfoTest.php (hunk 12 above)
Comment (keradus) [inline tests/ToolInfoTest.php]: then what about moving this one assertion to that existing method?

File: tests/ToolInfoTest.php [hunk 14]
```diff
@@ -21,13 +21,90 @@
...
+        $this->setExpectedException('LogicException');
+
+        $toolInfo->getComposerVersion();
+    }
+
+    private static function getPhpunitBin()
+    {
+        static $phpunitBin;
+        if (null === $phpunitBin) {
+            $phpunitBin = DIRECTORY_SEPARATOR.'..'.DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'bin'.DIRECTORY_SEPARATOR.'phpunit';
```
Comment (keradus) [inline tests/ToolInfoTest.php]: no need for cache in global state, this is just concatenation

File: src/ToolInfo.php (hunk 13 above)
Comment (Slamdunk) [inline src/ToolInfo.php]: The custom `bin-dir` is a good point, but:
1. if you are under Windows
1. and you use a custom `bin-dir`
there is no way to properly detect a composer installation since bins created by composer on Windows are not symlinks.

A step behide: is this test really necessary on Windows?
On Linux I usually create symlink in my `$PATH`, but on Windows linking is hard and users/softwares prefer to add another path to `$PATH` AFAIK.

Soooo, what should I do?

File: tests/ToolInfoTest.php (hunk 12 above)
Comment (Slamdunk) [inline tests/ToolInfoTest.php]: `testIsInstalledAsPhar` is a bit misleading: currently it tests that it is NOT installed by Phar, because `'phar://' === substr(__DIR__, 0, 7)` is not mockable.

The test would look like:
```php
public function testIsInstallAsPhar()
{
    $toolInfo = new ToolInfo(null);
    $this->assertFalse($toolInfo->isInstalledAsPhar());
    $this->assertFalse($toolInfo->isInstalledByComposer());
}
```

File: src/ToolInfo.php (hunk 13 above)
Comment (julienfalque) [inline src/ToolInfo.php]: The test works now so there is no reason to remove it :)

For the custom `bin-dir`, maybe we can check if the filename ends with `.bat` instead?

File: src/ToolInfo.php [hunk 15]
```diff
@@ -23,83 +23,114 @@
...
 
-            $result = dirname($script);
+            $composerInstalledJson = DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'installed.json';
+            // Windows real file is under `vendor/bin` because composer
+            // doesn't use symlinks on that OS
+            // On Linux instead the executabile is under `vendor/maintainer/package`
+            $this->composerInstalledFile = dirname(dirname($script)).$composerInstalledJson;
+            if (!file_exists($this->composerInstalledFile)) {
+                $this->composerInstalledFile = dirname(dirname(dirname($script))).$composerInstalledJson;
+            }
```
Comment (Slamdunk) [inline src/ToolInfo.php]: @julienfalque [bin-dir](https://getcomposer.org/doc/06-config.md#bin-dir) could be in any directory. The only way to know the exact position of `vendor` (and therefore installed.json) would be to read it, but this would lead to a strict coupling to composer package and its implementations.

Let me bring your attention to the fact that all thing code is here only to get the composer version (and it is installed through composer for cache).

What about removing everything and use https://github.com/Ocramius/PackageVersions ? @keradus ?

Comment (julienfalque) [inline src/ToolInfo.php]: Looks like a much better solution 👍

Comment (Slamdunk) [inline src/ToolInfo.php]: https://github.com/Ocramius/PackageVersions requires `~7.0` 😞

Comment (Slamdunk) [inline src/ToolInfo.php]: Ok, like Ocramius/PackageVersions I would to for `__DIR__` in replacement of `$_SERVER['SCRIPT_NAME']`, commits soon

**Comment by julienfalque** [comment]
Is it really important to rely on Composer to detect PHP-CS-Fixer version? Wouldn't it be easier to put it in a constant?

**Comment by Slamdunk** [comment]
Yes, if we want to expose `dev-master` commit

**Comment by Slamdunk** [comment]
What about now?

**Comment by julienfalque** [comment]
Now code is more simple but Composer detection will always fail for custom `bin-dir`.

**Comment by Slamdunk** [comment]
Why? The code now only looks for `ToolInfo.php` parents, no matter where you want your binaries.

File: src/ToolInfo.php [hunk 16]
```diff
@@ -73,31 +72,31 @@ public static function isInstalledByComposer()
         static $result;
 
         if (null === $result) {
-            $result = !self::isInstalledAsPhar() && file_exists(self::getScriptDir().self::COMPOSER_INSTALLED_FILE);
+            $result = !self::isInstalledAsPhar() && file_exists(self::getComposerInstalledFile());
```
Comment (julienfalque) [inline src/ToolInfo.php]: `file_exists()` can be replaced with `null !==` as the file is already checked in `getComposerInstalledFile()`.

File: src/ToolInfo.php [hunk 17]
```diff
@@ -73,31 +72,31 @@ public static function isInstalledByComposer()
...
-            if (is_link($script)) {
-                $linkTarget = readlink($script);
-
-                // If the link target is relative to the link
-                if (false === realpath($linkTarget)) {
-                    $linkTarget = dirname($script).'/'.$linkTarget;
+            $composerInstalledFile = '/composer/installed.json';
+            $checkedPaths = array(
+                // this project itself at ./src
+                __DIR__.'/../vendor'.$composerInstalledFile,
```
Comment (julienfalque) [inline src/ToolInfo.php]: Not sure we should check this one actually: the project does not require itself via Composer so I think the version won't be available in the json file.

File: src/ToolInfo.php [hunk 18]
```diff
@@ -73,31 +72,31 @@ public static function isInstalledByComposer()
...
+                // the project installed through composer at ./vendor/friendsofphp/php-cs-fixer/src
+                __DIR__.'/../../..'.$composerInstalledFile,
+            );
+
+            foreach ($checkedPaths as $path) {
+                $result = $path;
```
Comment (julienfalque) [inline src/ToolInfo.php]: This should be done in the `if`, otherwise the last checked file will be returned even if it doesn't exist.

File: src/ToolInfo.php (hunk 16 above)
Comment (Slamdunk) [inline src/ToolInfo.php]: What about now?

**Comment by julienfalque** [comment]
👍

File: src/ToolInfo.php [hunk 19]
```diff
@@ -73,33 +72,14 @@ public static function isInstalledByComposer()
...
-                }
-
-                $script = $linkTarget;
-            }
-
-            $result = dirname($script);
-        }
-
-        return $result;
+        return __DIR__.'/../../../composer/installed.json';
```
Comment (keradus) [inline src/ToolInfo.php]: I got lost. You check only this path, but from your initial comment:
> But composer installations can have only 2 result path, so just check them statically.

Comment (julienfalque) [inline src/ToolInfo.php]: See https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2637#discussion_r108693134

Comment (keradus) [inline src/ToolInfo.php]: thanks, makes sense

**Comment by keradus** [comment]
Thank you @Slamdunk.

**Comment by keradus** [comment]
info: merged into 2.1 instead of 2.2
